index.js

```diff
@@ -2,271 +2,169 @@
-
-    const dotIndex = feedId.lastIndexOf('.')
-    const b64part = feedId.substring(1, dotIndex)
-    const d = Buffer.from(b64part, 'base64')
-
-    return Buffer.concat([tf, d])
+  uri(input) {
+    let [type, format, ...data] = input.slice(6).split('/')
+    // *FIRE* WARNING ... are we URIEncoding our URIs?!
+    type = NAMED_TYPES[type]
+    if (type) {
+      format = type.formats[format]
+      if (format) {
+        return Buffer.concat([
+          type.code,
+          format.code,
+          Buffer.from(data.join('/'), 'base64'),
+        ])
+      }
+    }
+    throw new Error(`unknown type/format ${type.type}/${format.format}`)
   },
-
-  message(msgId) {
-    let tf
-    if (msgId.endsWith('.sha256')) tf = CLASSIC_MSG_TF
-    else if (msgId.endsWith('.ggmsg-v1')) tf = GABBYGR_MSG_TF
-    else if (msgId.endsWith('.bbmsg-v1')) tf = BENDYBT_MSG_TF
-    else throw new Error('Unknown msg ID: ' + msgId)
-
-    const dotIndex = msgId.lastIndexOf('.')
-    const b64part = msgId.substring(1, dotIndex)
-    const d = Buffer.from(b64part, 'base64')
-
-    return Buffer.concat([tf, d])
+  boolean(input) {
+    const d = input ? BOOL_TRUE : BOOL_FALSE
+    return Buffer.concat([BOOL_TF, d])
   },
+  classic(input, type, format) {
+    let data = input
+    if (type.sigil) data = data.slice(1)
+    if (format.suffix) data = data.replace(format.suffix, '')
 
-  blob(blobId) {
-    let tf
-    if (blobId.endsWith('.sha256')) tf = CLASSIC_BLOB_TF
-    else throw new Error('Unknown blob ID: ' + blobId)
-
-    const dotIndex = blobId.lastIndexOf('.')
-    const b64part = blobId.substring(1, dotIndex)
-    const d = Buffer.from(b64part, 'base64')
-
-    return Buffer.concat([tf, d])
+    return Buffer.concat([type.code, format.code, Buffer.from(data, 'base64')])
   },
-
-  box(boxedStr) {
-    if (boxedStr.endsWith('.box')) {
-      const b64part = boxedStr.substring(0, boxedStr.length - '.box'.length)
-      const d = Buffer.from(b64part, 'base64')
-      return Buffer.concat([BOX1_TF, d])
-    } else if (boxedStr.endsWith('.box2')) {
-      const b64part = boxedStr.substring(0, boxedStr.length - '.box2'.length)
-      const d = Buffer.from(b64part, 'base64')
-      return Buffer.concat([BOX2_TF, d])
-    } else throw new Error('Unknown boxed string: ' + boxedStr)
+  string(input) {
+    return Buffer.concat([Buffer.from([6, 0]), Buffer.from(input, 'utf8')])
   },
+}
 
-  signature(sig) {
-    const b64part = sig.substring(0, sig.length - '.sig.ed25519'.length)
-    const d = Buffer.from(b64part, 'base64')
-    return Buffer.concat([SIGNATURE_TF, d])
-  },
+function encode(input) {
+  /* cases we don't encode */
+  if (input === undefined || Buffer.isBuffer(input) || Number.isInteger(input))
+    return input
 
-  string(str) {
-    const d = Buffer.from(str, 'utf8')
-    return Buffer.concat([STRING_TF, d])
-  },
+  if (typeof input === 'string') {
+    if (input.startsWith('ssb://')) return encoder.uri(input)
```

Comment:
Not all SSB URIs have `//` some of them may be just `ssb:feed/sha256/`, see https://github.com/staltz/ssb-uri2/blob/5bb4c01ba8ba244c850c3e502ac635651d0dfebc/test/fixtures.js#L13-L15

Reply:
Good to know. Honestly there are multiple variants of this spec and I didn't know which to follow.
So yeah I'll pull the uri stuff out into a different pr + make it better tested

Reply:
removed